bamshuf.c

```diff
@@ -153,19 +283,146 @@ static int bamshuf(const char *fn, int n_files, const char *pre, int clevel,
+
+    if (fast) {
+        khash_t(bam_store) *stored = kh_init(bam_store);
+        khiter_t itr;
+        bam_list_t list;
+        int err = 0;
+
+        if (store_max < 2) store_max = 2;
+
+        if (create_bam_list(&list, store_max)) {
+            fprintf(stderr, "[collate[ ERROR: unable to create bam list.\n");
+            err = 1;
+            goto fast_fail;
+        }
+
+        while ((r = sam_read1(fp, h, list.items[list.index].b)) >= 0) {
+            int ret;
+            bam1_t *b = list.items[list.index].b;
+
+            if (!(b->core.flag & (BAM_FSECONDARY | BAM_FSUPPLEMENTARY))) {
+
+                itr = kh_get(bam_store, stored, bam_get_qname(b));
+
+                if (itr == kh_end(stored)) {
+                    // new read
+                    itr = kh_put(bam_store, stored, strdup(bam_get_qname(b)), &ret);
+
+                    if (ret > 0) { // okay to go ahead store it
+                        kh_value(stored, itr).bi = store_bam(&list);
+
+                        // see if the next one on the list needs to be written out
+                        if (write_bam_needed(&list)) {
+                            if (write_to_bin_file(list.items[list.index].b, cnt, fpt, fnt, h, n_files) < 0) {
+                                fprintf(stderr, "[collate] ERROR: could not write line.\n");
+                                err = 1;
+                                goto fast_fail;
+                            } else {
+                                mark_bam_as_written(&list);
+
+                                itr = kh_get(bam_store, stored, bam_get_qname(list.items[list.index].b));
+
+                                if (itr != kh_end(stored)) {
+                                    free((char *)kh_key(stored, itr));
+                                    kh_del(bam_store, stored, itr);
+                                } else {
+                                    fprintf(stderr, "[collate] ERROR: stored value not in hash.\n");
+                                    err = 1;
+                                    goto fast_fail;
+                                }
+                            }
+                        }
+                    } else if (ret == 0) {
+                        fprintf(stderr, "[collate] ERROR: value already in hash.\n");
+                        err = 1;
+                        goto fast_fail;
+                    } else {
+                        fprintf(stderr, "[collate] ERROR: unable to store in hash.\n");
+                        err = 1;
+                        goto fast_fail;
+                    }
+                } else { // we have a match
+                    // write out the reads in R1 R2 order
+                    bam1_t *r1, *r2;
+
+                    if (b->core.flag & BAM_FREAD1) {
+                        r1 = b;
+                        r2 = kh_value(stored, itr).bi->b;
+                    } else {
```

Comment:
Note that both BAM_FREAD1 and BAM_FREAD2 can be set at the same time.

Test needs to be `b->core.flag & (BAM_FREAD1 | BAM_FREAD2) == BAM_FREAD1` and similar for read2.

Reply:
Maybe, but the code is there to order the writing of a pair in R1 R2 order.  If one or both have BAM_FREAD1 and BAM_FREAD2 set what does that do to the order?  Does it have any meaning any more?

Reply:
Temlate-internal reads that are neither the first nor the last get READ1+READ2 set.  It's not a common thing to see though.

If there's filtering done elsewhere and we know only read 1 *or* read 2 can get to this code then this change will be irrelevant.